File: app/src/main/java/com/lasercyber/lws/ui/common/utils/SecretTapTracker.java
package com.lasercyber.lws.ui.common.utils;

import android.os.SystemClock;

/**
 * Detects consecutive taps within a short window (e.g. hidden settings entry).
 */
public final class SecretTapTracker {

    private static final int REQUIRED_TAPS = 5;
    /** Max idle time between consecutive taps before the count resets. */
    private static final long TAP_WINDOW_MS = 5000L;

    private int tapCount;
    private long resetAfterElapsed;

    /** @return {@code true} when the required number of taps was reached */
    public boolean registerTap() {
        long now = SystemClock.elapsedRealtime();
        if (now > resetAfterElapsed) {
            tapCount = 0;
        }
        tapCount++;
        resetAfterElapsed = now + TAP_WINDOW_MS;
        if (tapCount >= REQUIRED_TAPS) {
            tapCount = 0;
            return true;
        }
        return false;
    }

    public void reset() {
        tapCount = 0;
        resetAfterElapsed = 0L;
    }
}

File: app/lws_hmi/lib/platform/cloud/secret_tap_tracker.dart
/// Detects consecutive taps within a short window (lws-ui `SecretTapTracker`).
final class SecretTapTracker {
  SecretTapTracker({
    this.requiredTaps = 5,
    this.tapWindow = const Duration(seconds: 5),
  });

  final int requiredTaps;
  final Duration tapWindow;

  int _tapCount = 0;
  DateTime? _resetAfter;

  /// Returns `true` when [requiredTaps] were reached within [tapWindow].
  bool registerTap([DateTime? now]) {
    final t = now ?? DateTime.now();
    final deadline = _resetAfter;
    if (deadline == null || t.isAfter(deadline)) {
      _tapCount = 0;
    }
    _tapCount++;
    _resetAfter = t.add(tapWindow);
    if (_tapCount >= requiredTaps) {
      _tapCount = 0;
      return true;
    }
    return false;
  }

  void reset() {
    _tapCount = 0;
    _resetAfter = null;
  }
}
